File: Algorithms/src/962. Maximum Width Ramp/Code/solution.go
func maxWidthRamp(nums []int) int {
    n := len(nums)
    stack := []int{}
    
    // Step 1: Build a decreasing stack of indices
    for i := 0; i < n; i++ {
        if len(stack) == 0 || nums[stack[len(stack)-1]] > nums[i] {
            stack = append(stack, i)
        }
    }
    
    maxWidth := 0
    
    // Step 2: Traverse from the end and find maximum width ramp
    for j := n - 1; j >= 0; j-- {
        for len(stack) > 0 && nums[stack[len(stack)-1]] <= nums[j] {
            i := stack[len(stack)-1]
            stack = stack[:len(stack)-1]
            if j - i > maxWidth {
                maxWidth = j - i
            }
        }
    }
    
    return maxWidth
}

File: Algorithms/src/962. Maximum Width Ramp/Code/solution.java
class Solution {
    public int maxWidthRamp(int[] nums) {
        int n = nums.length;
        Stack<Integer> stack = new Stack<>();
        
        // Step 1: Build a decreasing stack of indices
        for (int i = 0; i < n; ++i) {
            if (stack.isEmpty() || nums[stack.peek()] > nums[i]) {
                stack.push(i);
            }
        }
        
        int maxWidth = 0;
        
        // Step 2: Traverse from the end and find maximum width ramp
        for (int j = n - 1; j >= 0; --j) {
            while (!stack.isEmpty() && nums[stack.peek()] <= nums[j]) {
                maxWidth = Math.max(maxWidth, j - stack.pop());
            }
        }
        
        return maxWidth;
    }
}